doc/user/content/sql/create-sink/kafka.md

```diff
@@ -69,33 +70,253 @@ Field                | Value  | Description
+[`uint8`]                    | `{"type": "fixed", "size": 8, "name": "com.materialize.sink.uint8"}`
+[`timestamp`]                | `{"type": "long", "logicalType: "timestamp-micros"}`
+[`timestamp with time zone`] | `{"type": "long", "logicalType: "timestamp-micros"}`
+[Arrays]                     | `{"type": "array", "items": ...}`
 
-To create a sink that uses the standard key-value convention to support inserts, updates, and deletes in the sink topic, you can use `ENVELOPE UPSERT`:
+### JSON
 
-```sql
-CREATE SINK avro_sink
-  FROM <source, table or mview>
-  INTO KAFKA CONNECTION kafka_connection (TOPIC 'test_avro_topic')
-  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_connection
-  ENVELOPE UPSERT
-  WITH (SIZE = '3xsmall');
+<b>Syntax:</b> `FORMAT JSON`
+
+When using the Avro format, the value of each Kafka message is a JSON object
+containing a field for each column of the sink's source relation. The names and
+ordering of the fields in the record match the names and ordering of the columns
+in the relation.
+
+If the `KEY` option is specified, the key of each Kafka message is a JSON
+object containing a field for each key column, in the same order and with the
+same names.
+
+SQL values are converted to JSON values according to the following rules:
+
+  * [`boolean`] values are converted to `true` or `false`.
+  * [`smallint`], [`integer`], [`bigint`], [`uint2`], [`uint4`], [`uint8`],
+    [`real`], and [`double precision`] values are converted to JSON numbers.
+  * [`numeric`] values are converted to a JSON string containing the
+    decimal representation of the number.
+  * [`record`] types are converted to JSON objects. The names and ordering of
+    the fields in the object match the names and ordering of the fields in the
+    record.
+  * [`list`] values are converted to JSON arrays.
+  * [array][arrays] values are converted to JSON arrays. Multidimensional arrays
+    are flattened into a single dimension following row-major order.
+  * [`timestamp`] and [`timestamp with time zone`] values are converted to JSON
+    strings containing the number of milliseconds since the Unix epoch.
+  * All other types are cast to [`text`] and then converted to JSON strings.
+
+## Envelopes
+
+The sink's envelope determines how inserts, updates, and delete events are
+mapped to Kafka messages.
+
+### Upsert
+
+<b>Syntax:</b> `ENVELOPE UPSERT`
+
+The upsert envelope:
+
+  * Emits insertion and update events without additional decoration.
+  * Converts deletion events into messages with a `null` value (often called
+    "tombstones").
+  * Requires that you specify a unique key for the sink's source
+    relation using the `KEY` option.
+
+Consider using the upsert envelope if:
+
+  * You need to follow standard Kafka conventions for upsert semantics.
+  * You want to enable key-based compaction on the sink's Kafka topic while
+    retaining the most recent value for each key.
+
+#### Key selection
```

Comment:
The content in this section is very troubleshooting-focused, so we're again mixing and matching reference documentation and guides, but we can figure out how to move content around later.